### PR TITLE
Basic Elevators

### PR DESCRIPTION
**Pull request by DrDuckedGoose**
## About The Pull Request

Adds basic elevators for mappers to use on multi-z maps.

It's pretty neat, it's animated. The elevator, and it's contents, shrink and grow as it traverses levels.
It can crush people and throw them out of the way, pretty sick.
Creation is pretty easy too. You physically build it in the map file, you don't have to mess with spawners.

## Why It's Good For The Game

Adds an alternative to the god-awful current solution.
I need it for AbyssStation.

## Testing Photographs and Procedure
<details>
<summary>Screenshots&Videos</summary>

My internet sucks right now, I couldn't get any video stuff.

image
image
image
image
image

</details>

- [x] Make TGUI for elevator interface
- [x] Flesh out move blacklist
- [x] Fix doors abusing the move process with sleep()

## Changelog
:cl: @DrDuckedGoose Code, @itsmeow UI, @qwertyquerty Sound
add: Added basic elevators, with control elements.
tweak: Replaces the elevator on multi-z debug with new presets
add: Added elevator UI theme, with fancy elevator buttons.
/:cl:

### CODE REVIEW

**Comment by DrDuckedGoose** [comment]
I will remove the debug map from the map prefs when I mark it ready for review.

**Comment by EvilDragonfiend** [comment]
If elevator comes to the bottom and there's a person, do they get gibbed?
Does the elevator have a ceiling?
If some tiling is placed above the elevator, will it block the elevator?

**Comment by DrDuckedGoose** [comment]
> If elevator comes to the bottom and there's a person, do they get gibbed?
They recieve 15 brute and are thrown very far away.
> Does the elevator have a ceiling?
No.
> If some tiling is placed above the elevator, will it block the elevator?
Not at the moment, no. Should it?

**Comment by PowerfulBacon** [comment]
What happened to the working elevators already in the codebase (turbolifts)?

**Comment by DrDuckedGoose** [comment]
Not maintained. They don't spawn in properly or respond to prompts.

File: config/maps.txt [hunk 1]
```diff
@@ -63,5 +63,5 @@ map runtimestation
 endmap
 
 map multiz_debug
-	disabled
+	enabled
```
Comment (Ivniinvi) [inline config/maps.txt]: Debug map enabled
```suggestion
	disabled
```

**Comment by DrDuckedGoose** [comment]
Whoops.
Thanks mate.

File: code/modules/elevator/elevator_controller.dm [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+		elevator_groups[id] = list()
+	elevator_groups[id] |= EV
+	
+/datum/controller/subsystem/elevator_controller/proc/move_elevator(id, destination_z, calltime, force)
+	. = TRUE
+	elevator_group_timers[id] = addtimer(CALLBACK(src, .proc/finish_timer, id), calltime || 2 SECONDS, TIMER_STOPPABLE)
+	//Loop through group ID, to assure there isn't anything blocking us
+	var/crashing = FALSE
+	var/obj/structure/elevator_segment/S = elevator_groups[id][1]
+	if((abs(destination_z - S.z) > 1 || destination_z > S.z))
```
Comment (itsmeow) [inline code/modules/elevator/elevator_controller.dm]: all `.z` should be `.get_virtual_z_level()`

File: code/modules/elevator/elevator_segment.dm [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+	//Throw mobs out below us
+	for(var/mob/living/i in destination.contents)
+		//If it's a mob, throw it out of the way
+		if(z_destination < old_z_this)
+			var/turf/trg = get_edge_target_turf(i, pick(NORTH, EAST, SOUTH, WEST))
+			i.throw_at(trg, 8, 8)
+			i.Paralyze(8 SECONDS)
+			i.adjustBruteLoss(15)
+			playsound(i, 'sound/effects/blobattack.ogg', 40, TRUE)
+			playsound(i, 'sound/effects/splat.ogg', 50, TRUE)
```
Comment (itsmeow) [inline code/modules/elevator/elevator_segment.dm:91]: use the squash component, it'll be funny

File: code/modules/elevator/elevator_segment.dm [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+	animate(target, transform = otransform, time = calltime / 2 || 1 SECONDS, flags = ANIMATION_PARALLEL)
+	//pixel adjustments
+	var/x_diff = SSelevator_controller.elevator_group_positions[id]["middle"]["x"] - target.x
+	var/y_diff = SSelevator_controller.elevator_group_positions[id]["middle"]["y"] - target.y
+	var/ox = target.pixel_x
+	var/oy = target.pixel_y
+	target.pixel_x = x_diff * (icon_size - icon_size * scale)
+	target.pixel_y = y_diff * (icon_size - icon_size * scale)
+	animate(target, pixel_x = ox, time = calltime / 2 || 1 SECONDS, flags = ANIMATION_PARALLEL)
+	animate(target, pixel_y = oy, time = calltime / 2 || 1 SECONDS, flags = ANIMATION_PARALLEL)
```
Comment (itsmeow) [inline code/modules/elevator/elevator_segment.dm:156]: this is going to break the shit out of any scaled mobs or items, we should either accept that and continue or not do it for mobs

Most important for dwarfism/giant mobs

**Review by itsmeow — CHANGES_REQUESTED** [review]

**Comment by itsmeow** [comment]
image
pressing the call button should not also open the UI for the outside type, only the inside should have buttons

**Comment by itsmeow** [comment]
image
somehow this didn't break my "resize", and for that I applaud you

File: code/modules/elevator/elevator_segment.dm (hunk 4 above)
Comment (itsmeow) [reply code/modules/elevator/elevator_segment.dm:156]: nevermind apparently

**Comment by itsmeow** [comment]
There's no door on the shaft if the elevator is not on that floor... this seems bad
image

**Comment by itsmeow** [comment]
https://user-images.githubusercontent.com/10366817/223573643-9a344394-8bfe-4984-abda-99319239e4e4.mp4

**Comment by itsmeow** [comment]
Make it play `sound/effects/turbolift/elevatormusic.ogg` while it moves. Also it should shake the screen slightly when moving and have a rumble and an impact sound when it lands

**Comment by Tsar-Salat** [comment]
> Make it play `sound/effects/turbolift/elevatormusic.ogg` while it moves. Also it should shake the screen slightly when moving and have a rumble and an impact sound when it lands

If doing this, I recommend getting rid of the door sounds and bolt noises, maybe use a resprited firelock instead?

Otherwise they will take away from the elevator effect you may be going for (Personally they just sound kinda janky here)

**Review by itsmeow — CHANGES_REQUESTED** [review]
Delete turbolift code and icons from the game

**Comment by DrDuckedGoose** [comment]
Done-zo

**Comment by itsmeow** [comment]
The music doesn't seem to trigger when moving from floor 3 to floor 1 on multiz debug

**Comment by itsmeow** [comment]
If possible... It would be very cool if the music just played anytime you were *in* the elevator, since the transit time is pretty short and the tracks are pretty long

File: code/modules/elevator/elevator_segment.dm [hunk 5]
```diff
@@ -8,6 +8,9 @@
 	var/turf/base_turf
 	///List of things we refuse to transport
 	var/static/list/move_blacklist = typecacheof(list(/atom/movable/lighting_object, /obj/structure/cable, /obj/structure/disposalpipe, /obj/machinery/atmospherics/pipe))
+	///List of elevator music files
+	var/list/music_files = list('sound/effects/turbolift/elevatormusic.ogg', 'sound/effects/turbolift/elevator_loop.ogg')
```
Comment (itsmeow) [inline code/modules/elevator/elevator_segment.dm]: one last thing, initializing typecaches and lists or empty lists outside of `Initialize` is bad for performance, you should move all list assignments and typecache inits to that. same applies to the other new code. subsystems I think are an exception case

**Review by itsmeow — CHANGES_REQUESTED** [review]

**Comment by itsmeow** [comment]
also mapping errors, appears there were two leftover turbolift things

**Comment by itsmeow** [comment]
Initialize must call parent

**Comment by itsmeow** [comment]
The second (new) ambiance is too loud

Also, if you can, pan the sound spatially so it sounds like it's coming from the right

File: code/modules/elevator/elevator_indicator.dm [hunk 6]
```diff
@@ -0,0 +1,35 @@
+/obj/machinery/elevator_indicator
+	name = "floor indicator"
+	icon = 'icons/obj/elevator.dmi'
```
Comment (itsmeow) [inline code/modules/elevator/elevator_indicator.dm:3]: this needs a desc, and it should probably also list the text on it when examined

**Review by itsmeow — CHANGES_REQUESTED** [review]

File: code/modules/elevator/elevator_indicator.dm (hunk 6 above)
Comment (itsmeow) [reply code/modules/elevator/elevator_indicator.dm:3]: control panel also needs a desc

**Comment by DrDuckedGoose** [comment]
Each music track now has it's own volume define. I don't think I can play the sound file in a specific ear, however.
I'm not aware of anyway to change the sound channel stuff like that.

**Comment by itsmeow** [comment]
the `sound()` proc has args for x y and z, you just put x as like 5 and it plays as if it was 5 to your right

**Review by itsmeow — APPROVED** [review]
Yeah I like this a lot, it works well ingame from my testing.

**Comment by itsmeow** [comment]
Needs new proc refs

**Comment by itsmeow** [comment]
that was fast

**Comment by DrDuckedGoose** [comment]
Did it in -2 minutes

**Review by itsmeow — APPROVED** [review]

**Review by itsmeow — APPROVED** [review]

**Pull request merged**